src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -675,6 +675,14 @@ public String resolveIndexRouting(@Nullable String routing, String aliasOrIndex)
 
             aliasesOrIndices = convertFromWildcards(aliasesOrIndices, indicesOptions);
         }
+
+        if (isAllIndices(aliasesOrIndices)) {
+            if (!indicesOptions.allowNoIndices()) {
+                throw new IndexMissingException(new Index("_all"));
+            }
+            else return Strings.EMPTY_ARRAY;
+        }
+
```

Comment:
ah I see, that's because you pass in `null` as second argument. I guess in practice most of the apis fall into the expand wildcards case, that is why we have never seen this problem in reality. 

I think you should just do:

```
if (aliasesOrIndices == null || aliasesOrIndices.length == 0)
```
then instead of
```
if (isAllIndices(aliasesOrIndices)) 
```

and adapt the error message. This doesn't have to do with the `_all` expression, which is treated as a special wildcard.

Note that we were previously missing to check `allowNoIndices` when `aliasesOrIndices` is empty... good test and good change! would be nice to figure out what the impact of this is, I think it might deserve the breaking label.... actually, maybe I would even pull out this change and get it in separately. What do you think?